metrics/micrometer/src/main/java/com/datastax/oss/driver/internal/metrics/micrometer/MicrometerMetricUpdater.java

```diff
@@ -100,7 +100,10 @@ protected void initializeGauge(
           m -> {
             MetricId id = getMetricId(m);
             Iterable<Tag> tags = MicrometerTags.toMicrometerTags(id.getTags());
-            return Gauge.builder(id.getName(), supplier).tags(tags).register(registry);
+            return Gauge.builder(id.getName(), supplier)
+                .strongReference(false)
```

Comment:
This is really clever, but I wonder if we aren't masking a broader problem: 

When a session is closed, ideally, I'd like all metrics (session- and node-level ones) that were registered by it to be explicitly de-registered. This is not the case currently, and that's imho the real cause why the memory leak is happening in the first place. 

Explicitly de-registering metrics is especially welcome if the sessions are sharing the same registry, which is the case by default when using Micrometer. If the registry is not shared, I bet there is no memory leak, but still, it'd be more correct to de-register everything.

My problem with the fix proposed here is that it will fix the issue for Micrometer gauges, but what if MicroProfile or Dropwizard change something in the future that would cause yet another strong reference to be retained by a metric?

I wonder if we shouldn't investigate introducing a new method `void deregisterMetrics()` in `MetricUpdater`.

Then we could modify `DefaultSession` and add this to `DefaultSession.close()` and `forceClose()`:

```java
metricUpdater.deregisterMetrics();
metadataManager
    .getMetadata()
    .getNodes()
    .values()
    .forEach(n -> ((DefaultNode) n).getMetricUpdater().deregisterMetrics());
```

The implementation for `MicrometerMetricUpdater` should be trivial (and for other impls too, I guess):

```java
  @Override
  public void deregisterMetrics() {
    metrics.values().forEach(registry::remove);
  }
```

But obviously, the devil is likely hiding in the details :-)

I'd be curious to hear what others think.

Reply:
Ooooooh, I _very_ much like this!  I _vastly_ prefer this to the current proposed fix; it's _far_ more explicit about what's going on and doesn't require complicated comments to explain what's going on.

FWIW I did [mention](https://issues.apache.org/jira/browse/CASSANDRA-19457?focusedCommentId=17824523&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17824523) some initial testing I'd done to try to determine the scope of this issue.  That testing seemed to indicate that the other metrics impls might be subject to similar issues but I'm now somewhat suspicious of those results, especially since the cause was so clearly tracked down to an issue with Micrometer.  Regardless, we should (a) re-confirm the current state of MicroProfile + DropWizard behaviour and (b) confirm that your proposed fix addresses any issues there (which it absolutely will).

@SiyaoIsHiding I very much agree with @adutra that something in this vein is a better way to implement metric lifetimes than what we have now.